File: 652.find-duplicate-subtrees.py
#
# @lc app=leetcode id=652 lang=python3
#
# [652] Find Duplicate Subtrees
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


from collections import deque
from typing import List, Optional


class Solution:
    '''
    Use a hash set to store the previously unseen nodes. As the root would never be a subtree, we should use postorder traversal to visit the tree, so that more simple tree structures are compared first. Each new node will be compared with all items in hash set and decide if it is a subtree. If not, add to the hash set. Tricky part in the algorithm: we need to serialize a subtree into string because we cannot handle that in the hash set.
   
    '''

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        # Breadth-first search
        # include null nodes as well!
        dq = deque()
        dq.append(root)
        bfs = []
        while dq:
            node = dq.popleft()
            if node:
                bfs.append(str(node.val))
                dq.append(node.left)
                dq.append(node.right)
            else:
                bfs.append("null")
        return ','.join(bfs)
                
    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode

        start with the first entry as root, then put each node into the queue. Use left or not left to define whether it is the left or right child currently pop from the queue when it is left        
        """

        if data == "null":
            return
        else:
            strings = data.split(",")
            root = TreeNode(int(strings[0]))
            dq = deque([root])
            left = True
            
            for string in strings[1:]:
                if string != "null":
                    node = TreeNode(int(string))
                    dq.append(node)
                else:
                    node = None
                if left:
                    parent = dq.popleft()
                    parent.left = node
                else:
                    parent.right = node
                left = not left
            return root
 
    def dfs(self, root):
        if root.left:
            self.dfs(root.left)
        if root.right:
            self.dfs(root.right)
        root_string = self.serialize(root)
        if root_string not in self.hash_set:
            self.hash_set.add(root_string)
        else:
            self.result.add(root_string)
    
    def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
        self.hash_set = set()
        self.result = set()
        self.dfs(root)
        res = []
        for node_string in self.result:
            res.append(self.deserialize(node_string))
        return res
# @lc code=end

File: 652.find-duplicate-subtrees.cs
/*
 * @lc app=leetcode id=652 lang=csharp
 *
 * [652] Find Duplicate Subtrees
 */

// @lc code=start

using System.Collections.Generic;
using System;
using System.Linq;

//Definition for a binary tree node.
//  public class TreeNode 
//  {
//     public int val;
//     public TreeNode left;
//     public TreeNode right;
//     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) 
//     {
//         this.val = val;
//         this.left = left;
//         this.right = right;
//     }
//   }
 
public class Solution 
{
    HashSet<string> seen;
    HashSet<string> result;
    public string serializeBinaryTree(TreeNode node)
    {
        Queue<TreeNode> queue = new Queue<TreeNode>();
        List<string> bfs = new List<string>();
        queue.Enqueue(node);
        while (queue.TryDequeue(out TreeNode node1))
        {
            if (node1 == null)
            {
                bfs.Add("null");
            }
            else
            {
                bfs.Add(node1.val.ToString());
                queue.Enqueue(node1.left);
                queue.Enqueue(node1.right);
            }
        }
        return string.Join(",", bfs.ToArray());
    }

    public TreeNode deserializeBinaryTree(string data)
    {
        if (data.Equals("null", StringComparison.OrdinalIgnoreCase))
        {
            return null;
        }
        else
        {
            string[] strings = data.Split(",");
            TreeNode root = new TreeNode(Int32.Parse(strings[0]));
            Queue<TreeNode> queue = new Queue<TreeNode>();
            queue.Enqueue(root);
            bool left = true;
            TreeNode node = default;
            TreeNode parent = default;            

            for (int i = 1; i < strings.Length; i++)
            {
                if (!strings[i].Equals("null", StringComparison.OrdinalIgnoreCase))
                {
                    node = new TreeNode(Int32.Parse(strings[i]));
                    queue.Enqueue(node);
                }
                else
                {
                    node = null;
                }

                if (left)
                {
                    parent = queue.Dequeue();
                    parent.left = node;
                }
                else
                {
                    parent.right = node;
                }
                left = !left;
            }
            return root;
        }
    }

    public void dfs(TreeNode root)
    {
        if (root.left != null)
        {
            dfs(root.left);
        }
        if (root.right != null)
        {
            dfs(root.right);
        }
        string rootString= serializeBinaryTree(root);

        if (!seen.Contains(rootString))
        {
            seen.Add(rootString);
        }
        else
        {
            result.Add(rootString);
        }
    }

    public IList<TreeNode> FindDuplicateSubtrees(TreeNode root) 
    {
        seen = new HashSet<string>();
        result = new HashSet<string>();
        dfs(root);
        return result.Select(x => deserializeBinaryTree(x)).ToList();
    }
}
// @lc code=end

